tests/compiler-hooks/build.gradle.kts

```diff
@@ -0,0 +1,209 @@
+                    .build()
+              } else {
+                member
+              }
+            } else {
+              member
+            }
+          }
+        }
+        .build()
+  }
+}
+
+/**
+ * Adds a default value to data class fields that are not nullable.
+ */
+private class DefaultNullValuesHooks : DefaultApolloCompilerKotlinHooks() {
+  override val version = "0"
+
+  override fun postProcessFileSpec(fileSpec: FileSpec): FileSpec {
+    return fileSpec
+        .toBuilder()
+        .apply {
+          members.replaceAll { member ->
+            if (member is TypeSpec) {
+              member.addDefaultValueToNullableProperties()
+            } else {
+              member
+            }
+          }
+        }
+        .build()
+  }
+
+  private fun TypeSpec.addDefaultValueToNullableProperties(): TypeSpec {
+    return toBuilder()
+        .apply {
+          // Only care about data classes
+          if (modifiers.contains(KModifier.DATA)) {
+            primaryConstructor(
+                primaryConstructor!!.toBuilder()
+                    .apply {
+                      parameters.replaceAll { param ->
+                        if (param.type.isNullable) {
+                          param.toBuilder()
+                              .defaultValue(CodeBlock.of("null"))
+                              .build()
+                        } else {
+                          param
+                        }
+                      }
+                    }
+                    .build()
+            )
+          }
+
+          // Recurse on nested types
+          typeSpecs.replaceAll { typeSpec ->
+            typeSpec.addDefaultValueToNullableProperties()
+          }
+        }
+        .build()
+  }
+}
+
+/**
+ * Adds a super interface to models that expose __typename.
+ */
+private class TypeNameInterfaceHooks(private val interfaceName: String) : DefaultApolloCompilerKotlinHooks() {
+  override val version = "1"
+
+  override fun postProcessFileSpec(fileSpec: FileSpec): FileSpec {
+    return fileSpec
+        .toBuilder()
+        .apply {
+          members.replaceAll { member ->
+            if (member is TypeSpec) {
+              member.addSuperInterfaceOnType()
+            } else {
+              member
+            }
+          }
+        }
+        .build()
+  }
+
+  private fun TypeSpec.addSuperInterfaceOnType(): TypeSpec {
+    return toBuilder()
+        .apply {
+          var hasTypeName = false
+          propertySpecs.replaceAll { propertySpec ->
+            if (propertySpec.name == "__typename") {
+              hasTypeName = true
+              propertySpec.toBuilder()
+                  .addModifiers(KModifier.OVERRIDE)
+                  .build()
+            } else {
+              propertySpec
+            }
+          }
+          if (hasTypeName) {
+            // XXX addSuperinterface(ClassName.bestGuess(interfaceName)) doesn't compile for an unknown reason, but this is equivalent
```

Comment:
I have no explanation. Using `addSuperinterface` triggers:

```
Overload resolution ambiguity: 
public final fun addSuperinterface(superinterface: KClass<*>, delegate: CodeBlock = ...): TypeSpec.Builder defined in com.squareup.kotlinpoet.TypeSpec.Builder
public final fun addSuperinterface(superinterface: TypeName, delegate: CodeBlock = ...): TypeSpec.Builder defined in com.squareup.kotlinpoet.TypeSpec.Builder
```

but that doesn't make sense to me. Maybe an issue related to r8?

Reply:
That's because we relocate the stdlib and `KClass` is relocated so the `KClass` from the `addSuperinterface` method is a different one from the one in `build.gradle.kts`

All in all, I think it's ok to have users of this feature rely on the `external` plugin. This was we don't have to keep kotlinpoet and don't risk breaking existing consumers.

Reply:
Good catch about KClass!

Ok to make it work with the `external` plugin, I agree that it's acceptable for this somewhat advanced feature